components/script/dom/element.rs

```diff
@@ -2501,11 +2502,32 @@ impl ElementMethods for Element {
 
     // https://drafts.csswg.org/cssom-view/#dom-element-clientwidth
     fn ClientWidth(&self) -> i32 {
+        let owner_doc: &Document = &self.node.owner_doc();
+        if self.local_name().to_string() == "body" ||
+            matches!(owner_doc.quirks_mode(), QuirksMode::Quirks)
+        {
+            let window: &Window = owner_doc.window();
+            return window.window_size().initial_viewport.round().to_i32().width;
+        }
+
```

Comment:
Thanks for working on this.

The specification text here is:

> If the element is the [root element](https://w3c.github.io/csswg-drafts/css-display-4/#root-element) and the element’s [node document](https://dom.spec.whatwg.org/#concept-node-document) is not in [quirks mode](https://dom.spec.whatwg.org/#concept-document-quirks), or if the element is [the body element](https://html.spec.whatwg.org/multipage/dom.html#the-body-element-2) and the element’s node document is in quirks mode, return the [viewport](https://www.w3.org/TR/CSS21/visuren.html#x1) width excluding the size of a rendered scroll bar (if any).

It seems like this code always returns the initial viewport width when the document is in quirks mode.

Look around in this file for how the body element and root element are detected. Let's use the same approach everywhere.

Let's also move this check into `client_rect` where it adjusts the width and height of the returned rectangle so it can be shared.

Reply:
Thanks for the suggestions! I'll implement your feedback as soon as possible.

Reply:
Also, is the initial viewport width always the same as the viewport width in general, or are there some differences between the two?

Reply:
Yes, I believe that's the case. Do you mind also pasting the spec text and linking to it above your check? Thanks!

Reply:
Yeah sure!

Reply:
I also noticed that in the Window class constructor, the current_viewport size is initialized to zero. Is this the correct behavior or is it supposed to be initialized to the dimensions of window_size?